### PR TITLE
Adds a function to decode arraybuffers as their string representation.

### PR DESCRIPTION
**Pull request by AlexaDeWit**
I spent a lot of time playing around with different base64 encoding things and adding support for browser testing, and eventually came to the conclusion that the transport-formats of an array buffer are fundamentally a different concern from dealing with array buffers themselves. I've squashed my work in progress here, which now just adds support for decoding an array buffer back into a string representation of it, using the uint16 view.

The base64 and transport stuff will move to its own pursuit package probably later this weekend or next weekend.

### CODE REVIEW

**Comment by AlexaDeWit** [comment]
I think I identified the problem. I had falsely assumed utf-8 encoding rather that code-point strings of utf16. I'll make the corrections shortly.

**Comment by AlexaDeWit** [comment]
Removing the test vector for now, since we're encoding the array buffers as codepoints rather than ut8 byte representations, so the test vector won't be valid. Given that the isomorphism seems to be holding for array buffers, I suspect we're breaking down on base64 stuff.

**Comment by AlexaDeWit** [comment]
So one concern I have generally is that this base64 encoding when combined with buffer-iffying strings via the purescript methodology (UC2, subset of utf16) rather than handling utf8 may lead to unexpected results for users wishing to work with simple string -> base64 encoding.

Maybe the best ways to dissuade incorrect usage here may be to constraint the buffer types for base64 encoding further, and provide some String -> Base64 and Base64 -> Maybe String functions.

Although this starts to bring us into providing base64 stuff more than simply having it as a means to have a transfer format for the array buffers themselves. Thoughts, @jacereda ?

**Comment by AlexaDeWit** [comment]
We're still WIP here for now, as the browser side base64 doesn't work, since atob and btoa are only defined for the latin character subset. I'll be adding the compatibility necessities tomorrow.

File: src/Data/ArrayBuffer/ArrayBuffer.js [hunk 1]
```diff
@@ -6,31 +6,41 @@ exports.create = function(s) {
...
   return (new Uint8Array(s)).buffer;
-}
+};
 
 exports.fromIntArray = function(s) {
   return (new Uint8Array(s)).buffer;
-}
+};
 
 exports.fromString = function(s) {
```
Comment (jacereda) [inline src/Data/ArrayBuffer/ArrayBuffer.js:29]: Was there any issue in the previous implementation besides the missing semicolon?

File: src/Data/ArrayBuffer/ArrayBuffer.js [hunk 2]
```diff
@@ -6,31 +6,41 @@ exports.create = function(s) {
...
-}
+  var buf = new ArrayBuffer(s.length*2);
+  var bufView = new Uint16Array(buf);
+  for (var i=0, strLen=s.length; i<strLen; i++) {
+    bufView[i] = s.charCodeAt(i);
+  }
+ return buf;
+};
+
+exports.decodeToString = function(buffer) {
```
Comment (jacereda) [inline src/Data/ArrayBuffer/ArrayBuffer.js]: Should this return a `Maybe`? What if you pass a 1 byte array?

File: src/Data/ArrayBuffer/ArrayBuffer.js [hunk 3]
```diff
@@ -6,31 +6,41 @@ exports.create = function(s) {
...
+  const uintBuffer = new Uint16Array(buffer);
+  const reducer = function(accum, point) {
+      // use concat instead of es6 syntax for compatibility
+      return accum.concat([String.fromCharCode(point)]);
+  };
+  const points = uintBuffer.reduce(reducer, new Array());
+  return points.join("");
+};
```
Comment (jacereda) [inline src/Data/ArrayBuffer/ArrayBuffer.js]: I found this one-liner at https://stackoverflow.com/questions/6965107/converting-between-strings-and-arraybuffers 
```
function ab2str(buf) {
  return String.fromCharCode.apply(null, new Uint16Array(buf));
}```

File: .gitignore [hunk 4]
```diff
@@ -3,4 +3,4 @@ output/
 bower_components/
 node_modules/
 .psci
-.psci_modules/
+.psci_modules/
```
Comment (jacereda) [inline .gitignore]: Why should the newline be removed?

**Review by jacereda — CHANGES_REQUESTED** [review]
Hi, thanks for the PR and sorry for the delay...

File: src/Data/ArrayBuffer/ArrayBuffer.js (hunk 1 above)
Comment (AlexaDeWit) [reply src/Data/ArrayBuffer/ArrayBuffer.js:29]: Nah, I was just playing cleanup while I was in there.

File: src/Data/ArrayBuffer/ArrayBuffer.js (hunk 3 above)
Comment (AlexaDeWit) [reply src/Data/ArrayBuffer/ArrayBuffer.js]: I can take a look again in a bit, but I did previously have that and the isomorphism test was failing.  It was a week or more ago so I can't remember the reason for abdandoning that one-liner.

File: src/Data/ArrayBuffer/ArrayBuffer.js (hunk 2 above)
Comment (AlexaDeWit) [reply src/Data/ArrayBuffer/ArrayBuffer.js]: Good catch, yes. it absolutely should. An earlier revision did infact have a maybe, and I somehow forgot about that case as I went about simplifying.

File: .gitignore (hunk 4 above)
Comment (AlexaDeWit) [reply .gitignore]: Unnoticed editor mistake. Deleting it.

File: src/Data/ArrayBuffer/ArrayBuffer.js (hunk 3 above)
Comment (AlexaDeWit) [reply src/Data/ArrayBuffer/ArrayBuffer.js]: I reverted back to the one-liner, as it seems our isomorphism does indeed pass.

**Comment by jacereda** [comment]
Merged, thanks!

**Pull request merged**